### ISSUE TITLE
Base type info property types not resolved

### ISSUE THREAD

**Issue by fabiancook** [highsource/jsonix#135]
TLDR

Base type info doesn't work with properties that are complex types in another module
Solution provided (see pull request)

---------------------

Given we have a module:

```
{
  name: 'nz_org_tcf_v5_bif_party',
  dependencies: [],
  typeInfos: [
    {
      localName: 'PartyRole',
      typeName: {
        namespaceURI: 'http://www.tcf.org.nz/v5/bif/party',
        localPart: 'PartyRole'
      },
      propertyInfos: [
        {
          name: 'contactMedium',
          minOccurs: 0,
          collection: true,
          elementName: {
            localPart: 'ContactMedium'
          },
          typeInfo: '.ContactMedium'
        }
      ]
    }, {
      localName: 'ContactMedium',
      typeName: {
        namespaceURI: 'http://www.tcf.org.nz/v5/bif/party',
        localPart: 'PartyRole'
      },
      propertyInfos: [
        {
          name: 'email',
          elementName: {
            localPart: 'Email'
          },
          typeInfo: 'nz_org_tcf_v5_bif_types.Email'
        }
      ]
    }
  ]
}
```

Then we have another module that depends on the previous:

```
{
  name: 'nz_org_tcf_v5_bif_place',
  dependencies: ['nz_org_tcf_v5_bif_party'],
  typeInfos: [
    {
      localName: 'PlaceRole',
      typeName: {
        namespaceURI: 'http:\/\/www.tcf.org.nz\/v5\/bif\/place',
        localPart: 'PlaceRole'
      },
      baseTypeInfo: 'nz_org_tcf_v5_bif_party.PartyRole'
    }
  ]
}
```

We can clearly see that `PlaceRole` is extending `PartyRole`, and that `PartyRole` has a property that has the type of `ContactMedium`, I would expect this to work perfectly fine as I have defined `PartyRole` as having `baseTypeInfo` as `nz_org_tcf_v5_bif_party.PartyRole`

However this is not the case, what ends up happening is that I get an error saying `Type info [nz_org_tcf_v5_bif_place.ContactMedium] is not known in this context.`

I have experimented and succeeded in "fixing" this. 
Currently each property doesn't know about its parent class, which means it doesn't actually know that `.ContactMedium` relates to `nz_org_tcf_v5_bif_party.ContactMedium` and not `nz_org_tcf_v5_bif_place.ContactMedium`
By giving the property access to this information everything worked as expected, I did this by "attaching" the property to the type (so the property knows as well as the type)

**Comment by fabiancook** [comment]
An alternative is to resolve the typeInfo and baseTypeInfo names before they change modules, or even every time, that way we can rely on it always being resolved properly

**Comment by highsource** [comment]
Hi,

you're hiting #50, this is a long-known bug.
The workaround is to change the order of the mappings.
I've danced around this for a while so congratulations for finding the cause. :) Respect.

I don't quite like the idea of attaching properties to a certain class. Doesn't look quite good and does not address the core issue.

Which is: classes or type infos are built with context and module which are passed to the `build` method. The module is passed further on, *even in the case* when the structure changes module, for instance when we process parent properties - which are actually defined in a different module.

This happens here because of the inheritance of the properties. I'll think about the solution. Probably two-passe building. Not sure yet.

Best wishes,
Alexey

**Comment by fabiancook** [comment]
Do you think we should be building the entire base type before inheriting from it? This way the propertie's type info could be built before hand, then instead of building it with the dependent class it's done. 

I know the solution I originally came up with isn't to clean, had to get something working in the short term so I could continue using this project

**Comment by fabiancook** [comment]
Any updates on this?

**Comment by highsource** [comment]
Sorry, not yet. I'm a bit underwater at the moment.

**Comment by highsource** [comment]
Hi @fabiancook, could you please test the latest snapshot version?
I think I have fixed it.
